### PR TITLE
Working directory for executions strategies, class path fix for java project strats

### PR DESCRIPTION
**Pull request by jicata**
MsBuild/MsBuildLibrary compilers' processes take too long to .Kill() which results in System.IO.IOException when ExecutionStrategy tries to SafeDelete() its working directory - https://github.com/SoftUni-Internal/suls-issues/issues/3002
Java Project strategies use wrong class path  - https://github.com/SoftUni-Internal/suls-issues/issues/3149

### CODE REVIEW

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 1]
```diff
@@ -20,6 +22,20 @@ public static string CreateTempDirectory()
             }
         }
 
+        public static string CreateTempDirectoryForExecutionStrategy()
+        {
+            while (true)
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: I might say I am a little concerned about while(true)

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 2]
```diff
@@ -30,5 +46,32 @@ public static void SafeDeleteDirectory(string path, bool recursive = false)
                 Directory.Delete(path, recursive);
             }
         }
+
+        public static void DeleteExecutionStrategiesWorkingDirectories()
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: `DeleteExecutionStrategyWorkingDirectories` - by convention we tend to pluralize only last word

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 3]
```diff
@@ -30,5 +46,32 @@ public static void SafeDeleteDirectory(string path, bool recursive = false)
...
+        {
+            var directoryPaths = Directory.GetDirectories(GlobalConstants.ExecutionStrategiesWorkingDirectoryPath);
+            foreach (var dirPath in directoryPaths)
+            {
+                var dir = new DirectoryInfo(dirPath);
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: use `directory` and `directoryPath`

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 4]
```diff
@@ -30,5 +46,32 @@ public static void SafeDeleteDirectory(string path, bool recursive = false)
...
+                if (dir.Exists && dir.CreationTime < DateTime.Now.AddHours(-1))
+                {
+                    var isDeleted = false;
+                    var retryCount = 0;
+                    while (!isDeleted && retryCount <= 3)
+                    {
+                        try
+                        {
+                            SafeDeleteDirectory(dirPath, true);
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: `Save...`

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 5]
```diff
@@ -30,5 +46,32 @@ public static void SafeDeleteDirectory(string path, bool recursive = false)
...
+                            isDeleted = true;
+                        }
+                        catch
+                        {
+                            Thread.Sleep(1000);
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: 1000 in variable (the name of the variable is good to content the units)

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs [hunk 6]
```diff
@@ -279,8 +280,20 @@ public ActionResult RegisterJobForCleaningSubmissionsForProcessingTable()
         public void CleanSubmissionsForProcessing()
         {
             this.Data.Context.SubmissionsForProcessing
-                .Where(s => s.Processed && s.Processing == false)
+                .Where(s => s.Processed && !s.Processing)
                 .Delete();
         }
+
+        // TODO: Remove this method after the job is registered
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs]: Please, make sure you delete that code soon enough

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs [hunk 7]
```diff
@@ -279,8 +280,20 @@ public ActionResult RegisterJobForCleaningSubmissionsForProcessingTable()
...
+        public ActionResult RegisterJobForDeletingLeftOverFilesInTemp()
+        {
+            var cron = "0 1 * * *";
+            this.backgroundJobs.AddOrUpdateRecurringJob(
+                "DeleteLeftOverFoldersInTemp",
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs]: IMO `DeleteLeftOverFoldersInTempFolder`

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs [hunk 8]
```diff
@@ -279,8 +280,20 @@ public ActionResult RegisterJobForCleaningSubmissionsForProcessingTable()
         public void CleanSubmissionsForProcessing()
         {
             this.Data.Context.SubmissionsForProcessing
-                .Where(s => s.Processed && s.Processing == false)
+                .Where(s => s.Processed && !s.Processing)
                 .Delete();
         }
+
+        // TODO: Remove this method after the job is registered
+        public ActionResult RegisterJobForDeletingLeftOverFilesInTemp()
```
Comment (YanaSlavcheva) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ParticipantsController.cs]: IMO `RegisterJobForDeletingLeftOverFilesInTempFolder`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/ExecutionStrategy.cs [hunk 9]
```diff
@@ -19,14 +19,27 @@ public abstract class ExecutionStrategy : IExecutionStrategy
...
         public ExecutionResult SafeExecute(ExecutionContext executionContext)
         {
-            this.WorkingDirectory = DirectoryHelpers.CreateTempDirectory();        
+            Exception innerException = null;
+            this.WorkingDirectory = DirectoryHelpers.CreateTempDirectoryForExecutionStrategy();
             try
             {
                 return this.Execute(executionContext);
             }
+            catch (Exception e)
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/ExecutionStrategy.cs]: we use `ex` for exception, I believe

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/ExecutionStrategy.cs [hunk 10]
```diff
@@ -19,14 +19,27 @@ public abstract class ExecutionStrategy : IExecutionStrategy
...
+                throw;
+            }
             finally
             {
-                DirectoryHelpers.SafeDeleteDirectory(this.WorkingDirectory, true);
+                try
+                {
+                    DirectoryHelpers.SafeDeleteDirectory(this.WorkingDirectory, true);
+                }
+                catch (Exception e)
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/ExecutionStrategy.cs]: `ex`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaSpringAndHibernateProjectExecutionStrategy.cs [hunk 11]
```diff
@@ -37,15 +37,14 @@ public class JavaSpringAndHibernateProjectExecutionStrategy : JavaProjectTestsEx
...
             string mavenPath)
-            : base(javaExecutablePath, getCompilerPathFunc, javaLibsPath)
+            : base(javaExecutablePath, getCompilerPathFunc, javaLibrariesPath)
         {
             this.MavenPath = mavenPath;
-            this.ClassPath = $"-cp {this.JavaLibsPath}*;{this.WorkingDirectory}\\target\\* ";
         }
 
-        // GroupId - > ArtifactId,Version
+        // GroupId - > ArtifactId, Version
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaSpringAndHibernateProjectExecutionStrategy.cs]: ???

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs [hunk 12]
```diff
@@ -33,28 +33,24 @@ public class JavaUnitTestsExecutionStrategy : JavaZipFileCompileExecuteAndCheckE
...
         }
 
-        protected string JavaLibsPath { get; }
+        protected string JavaLibrariesPath { get; }
 
-        protected string ClassPath { get; set; }
-
-        protected string JUnitTestRunnerSourceFilePath { get; }
+        protected string JUnitTestRunnerSourceFilePath
+            => $"{this.WorkingDirectory}\\{JUnitRunnerClassName}{GlobalConstants.JavaSourceFileExtension}";
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaUnitTestsExecutionStrategy.cs]: `=>` on upper row

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs (hunk 1 above)
Comment (jicata) [reply Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: It's a measure against the randomness of the Path.GetRandomFileName() as it has (an incredibly low but still existent) chance to create a path that already exists.

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs (hunk 4 above)
Comment (jicata) [reply Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: `Safe` as in "safely" (adverb). Save cannot be used as imperative/verb doesn't make sense being used before another verb.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaSpringAndHibernateProjectExecutionStrategy.cs (hunk 11 above)
Comment (jicata) [reply Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaSpringAndHibernateProjectExecutionStrategy.cs]: Simply an annotation for what the grouping of the collection is, as Tuples are a little bit obfuscating Maven packages have ArtifactId and Version as their primary attribute properties.

Comment (YanaSlavcheva) [reply Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JavaSpringAndHibernateProjectExecutionStrategy.cs]: Ok, but I did had to think a while until I found out what you mean. Maybe it should be explained a little better. In a bit more descriptive way.

File: Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs [hunk 13]
```diff
@@ -50,23 +56,24 @@ public static void SafeDeleteDirectory(string path, bool recursive = false)
         public static void DeleteExecutionStrategiesWorkingDirectories()
         {
             var directoryPaths = Directory.GetDirectories(GlobalConstants.ExecutionStrategiesWorkingDirectoryPath);
-            foreach (var dirPath in directoryPaths)
+            const int threadSleepMilliseconds = 1000;
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Common/Extensions/DirectoryHelpers.cs]: you'd better place const on top of the file IMO

**Pull request merged**